feat(build): Another stab at having rollup tree-shaking in our build.

nosachamos opened the pull request:
To test this out, run:

`npm run serve.prod.rollup.exp`

Which will perform a regular prod build, but also tree-shake the app.js bundle with rollup. 

It will not tree-shake rxjs, but it shouldn't really matter given that only used operators should be imported, and the core content (Observables, Subject, etc) are used by most operators and apps. 

This is largely based on the work done in the rollup branch, but I did a few small changes like abandoning the idea of using rxjs-es (given that it is no longer published) and adding the commonsjs rollup support.

As of right now, the sample app that comes with the seed can be built without any changes or workarounds.

Preliminary results for the size of the generated app.js bundles look like this:

**serve.prod**
   app.js size: 1MB

**serve.prod.exp**
   app.js size: 623 Kb

**serve.prod.rollup.exp**
   app.js size: 426 Kb


What do you guys think?

mgechev commented:
@nosachamos pretty awesome, thanks for the PR! Would you add a line in README.md, and rename the task to `build.prod.rollup.aot` (or leave `build.prod.rollup.exp` and create `build.prod.rollup.aot` as an alias). It'll be great if in the readme you mention briefly the pre-requirements for tree-shaking an app and that developers should be careful with non-static es2015 imports.

Thanks again! I'm really excited about this feature.

nosachamos commented:
@mgechev you got it! I'll update in a few mins.

SamVerschueren commented on tools/tasks/seed/build.bundles.app.rollup.exp.ts:
```diff
@@ -0,0 +1,43 @@
+export = (done: any) => {
+  rollup.rollup(config)
+    .then((bundle: any) => {
+      var result = bundle.generate({
```
`const`

SamVerschueren commented on tools/tasks/seed/build.js.prod.rollup.exp.ts:
```diff
@@ -0,0 +1,44 @@
+    ]
+  }, Config.TMP_DIR);
+  let src = [
+    Config.TOOLS_DIR + '/manual_typings/**/*.d.ts',
```
`join`

SamVerschueren commented on tools/tasks/seed/transpile.bundles.rollup.exp.ts:
```diff
@@ -0,0 +1,37 @@
+  let result = gulp.src(src)
+    .pipe(plugins.plumber())
+    .pipe(tsProject())
+    .once('error', function(e: any) {
```
fat-arrow

SamVerschueren commented on tools/tasks/seed/transpile.bundles.rollup.exp.ts:
```diff
@@ -0,0 +1,37 @@
+ */
+
+export = () => {
+  let tsProject = makeTsProject({
```
`const`

SamVerschueren commented on tools/tasks/seed/transpile.bundles.rollup.exp.ts:
```diff
@@ -0,0 +1,37 @@
+    noFallthroughCasesInSwitch: false
+  }, Config.TMP_DIR);
+  let src = [
+    join(Config.TMP_DIR, 'bundle.js')
```
`const`

SamVerschueren commented on tools/tasks/seed/transpile.bundles.rollup.exp.ts:
```diff
@@ -0,0 +1,37 @@
+  let src = [
+    join(Config.TMP_DIR, 'bundle.js')
+  ];
+  let result = gulp.src(src)
```
`const`

SamVerschueren commented on tools/tasks/seed/build.js.prod.rollup.exp.ts:
```diff
@@ -0,0 +1,44 @@
+    Config.TOOLS_DIR + '/manual_typings/**/*.d.ts',
+    join(Config.TMP_DIR, '**/*.ts')
+  ];
+  let result = gulp.src(src)
```
`const`

SamVerschueren commented on tools/tasks/seed/build.js.prod.rollup.exp.ts:
```diff
@@ -0,0 +1,44 @@
+  let result = gulp.src(src)
+    .pipe(plugins.plumber())
+    .pipe(tsProject())
+    .once('error', function(e: any) {
```
fat-arrow

SamVerschueren commented on tools/tasks/seed/build.js.prod.rollup.exp.ts:
```diff
@@ -0,0 +1,44 @@
+ */
+
+export = () => {
+  let tsProject = makeTsProject({
```
`const`

SamVerschueren commented on tools/tasks/seed/build.bundles.app.rollup.exp.ts:
```diff
@@ -0,0 +1,43 @@
+        }
+        done();
+      });
+    }, function (error: any) {
```
fat-arrow

SamVerschueren replied:
I'm also more in favour of `.catch`, but that might just be me :).

SamVerschueren commented on tools/tasks/seed/build.bundles.app.rollup.exp.ts:
```diff
@@ -0,0 +1,43 @@
+
+const nodeResolve = require('rollup-plugin-node-resolve');
+const commonjs = require('rollup-plugin-commonjs');
+const rollup = require('rollup');
```
`import`? Or not available Here?

nosachamos replied:
Not available, I didn't find type definitions for these. But if anyone knows where I can find the typing I can add them and use import instead. Since types info is not available I am forced to use require instead.

nosachamos commented:
@SamVerschueren Awesome review, I have implemented all your suggestions, they're all good, thanks!

@mgechev I renamed the task to `serve.prod.rollup.aot`, updated the README.md and added an alias for `serve.prod.aot` given that the readme mentions it but we only had `serve.prod.exp` instead. 

Let me know if I forgot anything. Thanks again!

SamVerschueren commented:
Nice work @nosachamos :tada:!

mgechev commented:
@nosachamos thanks for the PR! @SamVerschueren thanks for reviewing it!
